src/peer_connection.cpp

```diff
@@ -5560,7 +5557,8 @@ namespace libtorrent
 	void peer_connection::setup_send()
 	{
 		TORRENT_ASSERT(is_single_thread());
-		if (m_disconnecting) return;
+
+		if (m_disconnecting || m_send_buffer.empty()) return;
```

Comment:
I notice the request bandwidth call is just below here. Is there a risk that if we don't have any quota, and the send buffer is empty, we won't ever request send quota?

Reply:
There are numerous traces where `m_send_buffer` is being modified, but the only function to consider quota is `peer_connection::setup_send()`. 

The following methods are known to modify send buffer regardless of upload quota:
1. `peer_connection::fill_send_buffer()`
2. `peer_connection::send_buffer()`
3. `peer_connection::append_send_buffer()`
4. `peer_connection::append_const_send_buffer()`

The only case I can't justify is related to send barrier.